### PR TITLE
fix(networks): only instantiate cross-attention layers when with_cross_attention=True

### LINKED ISSUE #8845: TransformerBlock instantiates CrossAttentionBlock even when with_cross_attention=False
**Describe the bug**
Hi ,

I noticed that TransformerBlock always instantiates cross-attention layers even when with_cross_attention=False.

Current implementation:

```
self.with_cross_attention = with_cross_attention

self.norm_cross_attn = nn.LayerNorm(hidden_size)

self.cross_attn = CrossAttentionBlock(
    hidden_size=hidden_size,
    num_heads=num_heads,
    dropout_rate=dropout_rate,
    qkv_bias=qkv_bias,
    causal=False,
    use_flash_attention=use_flash_attention,
)
```

but in forward:

```
if self.with_cross_attention:
    x = x + self.cross_attn(self.norm_cross_attn(x), context=context)
```

So when with_cross_attention=False the cross-attention modules are never used but their parameters still appear ( registered)  in model.parameters()

**To Reproduce**
With self.with_cross_attention =False  , 
```
 
    for name, p in model.named_parameters():
        if p.grad is None:
            print("[NO GRAD] ", name)
```
[NO GRAD] encoder.vit.vit.blocks.1.norm_cross_attn.weight
[NO GRAD] encoder.vit.vit.blocks.1.norm_cross_attn.bias
[NO GRAD] encoder.vit.vit.blocks.1.cross_attn.out_proj.weight
[NO GRAD] encoder.vit.vit.blocks.1.cross_attn.out_proj.bias
[NO GRAD] encoder.vit.vit.blocks.1.cross_attn.to_q.weight
[NO GRAD] encoder.vit.vit.blocks.1.cross_attn.to_q.bias
[NO GRAD] encoder.vit.vit.blocks.1.cross_attn.to_k.weight
[NO GRAD] encoder.vit.vit.blocks.1.cross_attn.to_k.bias
[NO GRAD] encoder.vit.vit.blocks.1.cross_attn.to_v.weight
[NO GRAD] encoder.vit.vit.blocks.1.cross_attn.to_v.bias

**Expected behavior**
There should be no params registered for the cross attention module , if the transformer block doesnt use cross attention.

### PR DESCRIPTION
**Pull request by AlexanderSanin**
## Summary

Fixes #8845

`TransformerBlock` was unconditionally creating `norm_cross_attn` and `cross_attn` in `__init__` regardless of the `with_cross_attention` flag. This caused unused parameters to appear in `model.parameters()` / `model.named_parameters()`, increasing model size and producing confusing "no gradient" warnings during training.

**Root cause** — two layers created unconditionally:
```python
# always ran, even when with_cross_attention=False
self.norm_cross_attn = nn.LayerNorm(hidden_size)
self.cross_attn = CrossAttentionBlock(...)
```

**Fix** — guard instantiation with the same flag already used in `forward`:
```python
if with_cross_attention:
    self.norm_cross_attn = nn.LayerNorm(hidden_size)
    self.cross_attn = CrossAttentionBlock(...)
```

## Changes

- `monai/networks/blocks/transformerblock.py`: wrap cross-attention layer init in `if with_cross_attention:`
- `tests/networks/blocks/test_transformerblock.py`: add two regression tests verifying parameter presence/absence

## Test plan

- [x] `test_no_cross_attention_params_when_disabled` — asserts no `cross_attn`/`norm_cross_attn` params when `with_cross_attention=False`
- [x] `test_cross_attention_params_when_enabled` — asserts cross-attn params exist when `with_cross_attention=True`
- [x] `test_ill_arg` — existing validation tests unchanged
- [x] Existing parameterized `test_shape` tests cover both `with_cross_attention=True/False` forward passes

https://claude.ai/code/session_01LV2dy8NFh3smu9f2RfgFvs

### CODE REVIEW

**Comment by AlexanderSanin** [comment]
Hey @ericspod @aymuos15. Could you, please, have a look at this?

**Comment by AlexanderSanin** [comment]
The failing CI checks are caused by a transient upstream dependency regression in `cupy-cuda12x` and are not related to this PR's changes.

**Root cause:** `cupy-cuda12x` 14.1.0 (released between May 22 and May 26) added an unconditional `import pytest` at module load time in `cupy/testing/_random.py`. `pytest` is not installed in the target environments for several CI jobs (only in `requirements-dev.txt`), so any import of `monai.networks` (which pulls in `polygraphy.backend.common` → `polygraphy.util.util` → `cupy.testing._random`) triggers a `ModuleNotFoundError`.

**Evidence:**
- Last successful `tests` workflow on `dev` (commit `0a8d945`, run [#26286623163](https://github.com/Project-MONAI/MONAI/actions/runs/26286623163), May 22): installed `cupy-cuda12x-14.0.1` → all jobs green
- This PR (same base commit `0a8d945`, May 26): installs `cupy-cuda12x-14.1.0` → `full-dep`, `static-checks (pytype)` fail; `mypy` and `codeformat` cancelled during dep install timeout
- The companion PR #8872 (different changes entirely) shows the exact same failure pattern, confirming it is environment-driven

**Fix scope:** This is a separate issue from this PR. The cleanest mitigation would be to pin `cupy-cuda12x!=14.0.0,!=14.1.0` in `requirements-dev.txt` (or add `pytest` to the runtime-import path's environment). Happy to open a separate PR for that if maintainers agree.

**This PR's tests pass locally:**
```
test_no_cross_attention_params_when_disabled PASSED
test_cross_attention_params_when_enabled     PASSED
test_ill_arg                                 PASSED
test_access_attn_matrix                      PASSED
```

**Comment by AlexanderSanin** [comment]
## Status & Blockers

**All code changes are complete and tests pass locally (4/4).** Two things are blocking merge:

### 1. CI Failures (external regression — not our code)

The failing CI jobs (`full-dep (ubuntu-latest)`, `static-checks (pytype)`, `mypy`, `codeformat`) are caused by a **`cupy-cuda12x` 14.1.0 regression** released on May 26, 2026. That version introduced `import pytest` at module load time in `cupy/testing/_random.py`, which breaks any environment where `pytest` is not installed.

The failure chain is:
`monai.networks` → `trt_compiler` → `polygraphy.backend.common` → `polygraphy.util.util` → `cupy.testing._random` → `import pytest` → `ModuleNotFoundError`

This is **pre-existing on the `dev` branch** — the last fully green dev CI run (#26286623163) was on May 22 with cupy 14.0.1. All our PR-specific tests pass.

### 2. Awaiting Required Review

This PR needs at least one approving review from a code owner before it can be merged.

@KumoLiu @ericspod @Nic-Ma — could one of you take a look when you get a chance? The change is in `monai/networks/blocks/transformerblock.py` and guards cross-attention layer instantiation behind `if with_cross_attention:` so that models initialized with `with_cross_attention=False` don't allocate unused parameters. Thanks!

> **Note on backwards compatibility:** Models saved with the old code that had `with_cross_attention=False` will have `cross_attn.*` keys in their `state_dict`. Loading those checkpoints with the fixed code requires `strict=False`.

**Comment by ericspod** [comment]
> ## Status & Blockers
> 
> **All code changes are complete and tests pass locally (4/4).** Two things are blocking merge:
> ### 1. CI Failures (external regression — not our code)

I'm seeing the same issue and am working on a fix right now on a PR I hope to integrate soon.

**Comment by ericspod** [comment]
Hi @AlexanderSanin we have #8848 already in progress which will fix this issue. Once we finish it we'll merge once tests are fixed, I'm sorry but this PR looks redundant once we do that. Thanks for the contribution regardless.

**Comment by AlexanderSanin** [comment]
Thanks @ericspod — agreed. Now that #8848 is merged and resolves #8845, this PR is redundant, so I'll close it.

For what it's worth, the merged `nn.Identity()` approach is the better call: it keeps the `cross_attn` / `norm_cross_attn` attributes present so existing checkpoints still load, which sidesteps the backward-compatibility concern I'd flagged here (old `state_dict`s containing `cross_attn.*` keys). Thanks for the guidance and for the quick turnaround on the CI fix!
